aptos-move/framework/aptos-framework/sources/configs/Stake.move

```diff
@@ -0,0 +1,247 @@
+    }
+
+    /// Initiate by the validator info owner.
+    fun remove_validator(account: &signer) acquires ValidatorInfo, ValidatorSet {
+        let addr = Signer::address_of(account);
+        let validator_set = borrow_global_mut<ValidatorSet>(@CoreResources);
+
+        let (exist, index) = Vector::index_of(&validator_set.validators, &addr);
+        assert!(exist, 0);
+        on_new_epoch(validator_set);
+        Vector::swap_remove(&mut validator_set.validators, index);
+        assert!(Vector::length(&validator_set.validators) > 0, 0);
+    }
+
+    /// Triggers when validator set changes or after certain time period (in block prologue).
+    fun on_new_epoch(validator_set: &mut ValidatorSet) acquires ValidatorInfo {
+        let current_time = Timestamp::now_seconds();
+        assert!(validator_set.last_update_time_secs + MINIMUM_RECONFIG_PERIOD > current_time, 0);
+        validator_set.last_update_time_secs = current_time;
+        let i = 0;
+        let len = Vector::length(&validator_set.validators);
+        while (i < len) {
+            let addr = *Vector::borrow(&validator_set.validators, i);
+            update_stake_info(addr);
+            i = i + 1;
+        };
+        // Reconfiguration::reconfigure();
+        // do we remove validators without enough stakes?
+    }
+
+    /// Update individual validator's stake info
+    /// 1. distribute rewards to active/cooldown delegations
+    /// 2. process warmup, cooldown correspondingly
+    /// 3. update the current stake
+    fun update_stake_info(addr: address) acquires ValidatorInfo {
+        let stake_info = &mut borrow_global_mut<ValidatorInfo>(addr).stake_info;
+        distribute_reward( &mut stake_info.active);
+        distribute_reward( &mut stake_info.cooldown);
+        // move warmup to active
+        append(&mut stake_info.active, &mut stake_info.warmup);
+        // move cooldown to inactive
+        append(&mut stake_info.inactive, &mut stake_info.cooldown);
+        let current_stake = 0;
+        let i = 0;
+        let len = Vector::length(&stake_info.active);
+        while (i < len) {
+            current_stake = current_stake + TestCoin::value(&Vector::borrow(&stake_info.active, i).coin);
+            i = i + 1;
+        };
+        stake_info.current_stake = current_stake;
+    }
+
+    /// Mint the reward and add to the delegation based on some formula
+    fun distribute_reward(v: &mut vector<Delegation>) {
+        let i = 0;
+        let len = Vector::length(v);
+        while (i < len) {
+            let d = Vector::borrow_mut(v, i);
+            let reward = TestCoin::zero(); // mint some coins based on delegation, timestamp, maybe also totaly stakes
+            TestCoin::merge(&mut d.coin, reward);
+        };
+    }
+
+    fun append<T>(v1: &mut vector<T>, v2: &mut vector<T>) {
+        while (!Vector::is_empty(v2)) {
+            Vector::push_back(v1, Vector::pop_back(v2));
+        }
+    }
```

Comment:
maybe call this take_one since there is already a function called Vector::append that has a very different behavior

Reply:
it's the same behavior except that one requires v2: vector<T> instead of &mut vector<T>